File: python/src/aoc_2024_05.py
"""Advent of Code 2024 - Day 5."""

from itertools import combinations
from sys import stdin


def read_pages() -> tuple[set[tuple[int, int]], list[list[int]]]:
    """Read page orderings and updates."""
    orderings: set[tuple[int, int]] = set()
    updates: list[list[int]] = []
    in_orderings = True
    for line in (s.strip() for s in stdin.readlines()):
        if not line:
            in_orderings = False
            continue
        if in_orderings:
            a, b = [int(s) for s in line.split("|")]
            orderings.add((a, b))
        else:
            updates.append([int(s) for s in line.split(",")])
    return (orderings, updates)


def valid_middle(update: list[int], orderings: set[tuple[int, int]]) -> int:
    """Return middle page if update is in valid order, zero otherwise."""
    for x, y in combinations(update, 2):
        if (y, x) in orderings:
            return 0
    return update[len(update) // 2]


def fixed_middle(update: list[int], orderings: set[tuple[int, int]]) -> int:
    """Return middle page after fixing if the update was invalid, zero otherwise."""
    any_invalid = False
    for i, j in combinations(range(len(update)), 2):
        if (update[j], update[i]) in orderings:
            any_invalid = True
            update[i], update[j] = update[j], update[i]
    return update[len(update) // 2] if any_invalid else 0


if __name__ == "__main__":
    orderings, updates = read_pages()
    print(sum(valid_middle(u, orderings) for u in updates))
    print(sum(fixed_middle(u, orderings) for u in updates))

File: rust/src/bin/aoc_2024_05.rs
// Advent of Code 2024 - Day 5.

use itertools::Itertools;
use std::collections::HashSet;
use std::io;

fn read_pages() -> (HashSet<(i32, i32)>, Vec<Vec<i32>>) {
    let mut orderings: HashSet<(i32, i32)> = HashSet::new();
    let mut updates: Vec<Vec<i32>> = Vec::new();
    let mut in_orderings = true;
    for raw_line in io::stdin().lines() {
        let line = raw_line.unwrap();
        let line = line.trim();
        if line.is_empty() {
            assert!(in_orderings);
            in_orderings = false;
            continue;
        }
        if in_orderings {
            let mut iter = line.split('|');
            let a: i32 = iter.next().unwrap().parse().unwrap();
            let b: i32 = iter.next().unwrap().parse().unwrap();
            assert!(iter.next().is_none());
            orderings.insert((a, b));
        } else {
            updates.push(line.split(',').map(|s| s.parse().unwrap()).collect());
        }
    }
    (orderings, updates)
}

fn valid_middle(update: &[i32], orderings: &HashSet<(i32, i32)>) -> i32 {
    for v in update.iter().combinations(2) {
        let [i, j] = v[..] else {
            panic!("Expected 2-vector from combinations(2)")
        };
        if orderings.contains(&(*j, *i)) {
            return 0;
        }
    }
    update[update.len() / 2]
}

fn fixed_middle(update: &[i32], orderings: &HashSet<(i32, i32)>) -> i32 {
    let mut any_invalid = false;
    let mut u: Vec<i32> = Vec::from(update);
    for v in (0..u.len()).combinations(2) {
        let [i, j] = v[..] else {
            panic!("Expected 2-vector from combinations(2)")
        };
        if orderings.contains(&(u[j], u[i])) {
            any_invalid = true;
            u.swap(i, j);
        }
    }
    if any_invalid {
        u[u.len() / 2]
    } else {
        0
    }
}

fn main() {
    let (orderings, updates) = read_pages();
    println!(
        "{}",
        updates
            .iter()
            .map(|u| valid_middle(u, &orderings))
            .sum::<i32>()
    );
    println!(
        "{}",
        updates
            .iter()
            .map(|u| fixed_middle(u, &orderings))
            .sum::<i32>()
    );
}
